lib2/Ppmd8Tdecoder.c

```diff
@@ -0,0 +1,138 @@
+    InBuffer *in = reader->inBuffer;
+    int max_length = args->max_length;
+    args->finished = False;
+    PPMD_pthread_mutex_unlock(&mutex);
+
+    Bool escaped = False;
+    int i = 0;
+    int result = 0;
+    while (i < max_length ) {
+        Bool can_break = False;
+        PPMD_pthread_mutex_lock(&mutex);
+        if (in->size == in->pos) {
+            can_break = True;
+        }
+        if (args->out->size == args->out->pos) {
+            can_break = True;
+        }
+        PPMD_pthread_mutex_unlock(&mutex);
+        if (can_break) {
+            break;
+        }
+        PPMD_pthread_mutex_lock(&mutex);
+        int c = Ppmd8_DecodeSymbol(cPpmd8);
+        PPMD_pthread_mutex_unlock(&mutex);
+        if (escaped) {
+            escaped = False;
+            if (c == 0x01) { // escaped character
+                PPMD_pthread_mutex_lock(&mutex);
+                *((Byte *)args->out->dst + args->out->pos++) = c;
+                PPMD_pthread_mutex_unlock(&mutex);
+                i++;
+            } else if (c == 0x00) { // endmark
+                // eof
+                result = -1;
+                goto exit;
+            } else {
+                // failed
+                result = -2;
+                goto exit;
+            }
+        } else {
+            if (c != 0x01) { // ordinary data
+                PPMD_pthread_mutex_lock(&mutex);
+                *((Byte *)args->out->dst + args->out->pos++) = (Byte) c;
+                PPMD_pthread_mutex_unlock(&mutex);
+                i++;
+            } else { // enter escape sequence
+                escaped = True;
+            }
+        }
+    }
+    // when success return produced size
+    result = i;
+
+    exit:
+    PPMD_pthread_mutex_lock(&mutex);
+    args->result = result;
+    args->finished = True;
+    PPMD_pthread_mutex_unlock(&mutex);
+    return NULL;
+}
+
+int Ppmd8T_decode(CPpmd8 *cPpmd8, OutBuffer *out, int max_length, ppmd8_args *args) {
+    PPMD_pthread_mutex_lock(&mutex);
+    args->cPpmd8 = cPpmd8;
+    args->max_length = max_length;
+    args->out = out;
+    BufferReader *reader = (BufferReader *) cPpmd8->Stream.In;
+    InBuffer *in = reader->inBuffer;
+    args->in = in;
+    args->result = 0;
+    Bool exited = args->finished;
+    PPMD_pthread_mutex_unlock(&mutex);
+
+    if (exited) {
+        PPMD_pthread_t handle;
+        PPMD_pthread_create(&handle, NULL, Ppmd8T_decode_run, args);
+        PPMD_pthread_mutex_lock(&mutex);
+        args->handle = handle;
+        PPMD_pthread_mutex_unlock(&mutex);
+    } else {
+        PPMD_pthread_mutex_lock(&mutex);
+        if (in->pos < in->size) {
+            PPMD_pthread_cond_signal(&notEmpty);
+            PPMD_pthread_mutex_unlock(&mutex);
+        } else {
+            PPMD_pthread_mutex_unlock(&mutex);
+            return -2;  // error
+        }
+    }
+    while(True) {
+        PPMD_pthread_mutex_lock(&mutex);
+        if (PPMD_pthread_cond_wait1(&inEmpty, &mutex) == 0) {
```

Comment:
Here blocks main thread 1msec. Function name is not explains it.
It should be given a better name or should add comment.